tools/kagami/src/genesis.rs

```diff
@@ -241,10 +259,251 @@ fn generate_synthetic(
+    let bytes = fs::read(path)?;
+    Ok(serde_json::from_slice(bytes.as_slice())?)
+}
+
+impl SignArgs {
+    fn get_private_key(&self) -> Result<PrivateKey> {
+        let private_key_bytes = get_key_raw(&self.private_key_file, &self.private_key_string)?;
+        match private_key_bytes {
+            KeyStorage::FromFile(bytes) => {
+                PrivateKey::from_bytes(self.algorithm.0, bytes.as_slice()).wrap_err_with(|| {
+                    eyre!(
+                        "Failed to parse private key from bytes for algorithm `{}`",
+                        self.algorithm
+                    )
+                })
+            }
+            KeyStorage::FromCLI(hex) => {
+                PrivateKey::from_hex(self.algorithm.0, hex).wrap_err_with(|| {
+                    eyre!(
+                        "Failed to parse private key from hex for algorithm `{}`",
+                        self.algorithm
+                    )
+                })
+            }
+        }
+    }
+
+    fn get_public_key(&self) -> Result<PublicKey> {
+        if let Some(key) = &self.public_key_string {
+            return Ok(key.clone());
+        }
+        let public_key_bytes = get_key_raw(&self.public_key_file, &None)?;
+        match public_key_bytes {
+            KeyStorage::FromFile(bytes) => {
+                PublicKey::from_bytes(self.algorithm.0, bytes.as_slice()).wrap_err_with(|| {
+                    eyre!(
+                        "Failed to parse public key from bytes for algorithm `{}`",
+                        self.algorithm
+                    )
+                })
+            }
+            KeyStorage::FromCLI(multihash_string) => PublicKey::from_str(multihash_string)
+                .wrap_err_with(|| {
+                    eyre!(
+                        "Failed to deserialize public key from multihash string for algorithm `{}`",
+                        self.algorithm
+                    )
+                }),
+        }
+    }
+}
+
+impl<T: Write> RunArgs<T> for SignArgs {
+    fn run(self, writer: &mut BufWriter<T>) -> Outcome {
+        let key_pair = if let Some(path) = self.keypair_file {
+            read_keypair(path)?
+        } else {
+            let (public_key, private_key) = (self.get_public_key()?, self.get_private_key()?);
+            KeyPair::new(public_key, private_key)?
+        };
+
+        let genesis_block = RawGenesisBlock::from_path(&self.genesis_file)?;
+        let genesis_signature =
+            GenesisNetwork::new_genesis_signature(genesis_block, &self.chain_id, &key_pair);
+        let genesis_signature_string = genesis_signature.to_hex_string();
+
+        let encoded_genesis_signature = if self.scale {
+            genesis_signature_string.encode()
+        } else {
+            Vec::default()
+        };
+
+        let hex_genesis_config = if self.hex {
+            genesis_signature_string
+        } else {
+            String::default()
+        };
+
```

Comment:
Certainly a case to use `Option`s for stronger abstractions.

```rust
let scale_encoded_signature = self.scale.then(|| genesis_signature_string.encode());
```

And `hex_genesis_config`... It doesn't seem to be needed.